File: backend/app/schemas/reports.py
"""
Reports Schemas (Phase 8)
Pydantic schemas for reports API responses
"""
from pydantic import BaseModel, Field
from typing import List, Dict, Optional
from datetime import date, datetime
from decimal import Decimal


# ============================================================================
# Revenue Report Schemas
# ============================================================================

class RevenueByPeriod(BaseModel):
    """Revenue data point for chart"""
    period: str  # "2025-01-15" or "2025-01"
    revenue: float
    count: int  # number of transactions


class RevenueReportResponse(BaseModel):
    """Revenue report response"""
    total_revenue: float
    total_transactions: int
    average_transaction: float
    by_payment_method: Dict[str, float]  # {"cash": 5000, "transfer": 3000}
    by_stay_type: Dict[str, float]  # {"overnight": 8000, "temporary": 2000}
    by_period: List[RevenueByPeriod]  # Chart data
    start_date: date
    end_date: date


# ============================================================================
# Occupancy Report Schemas
# ============================================================================

class OccupancyByPeriod(BaseModel):
    """Occupancy data point for chart"""
    period: str  # "2025-01-15"
    occupancy_rate: float  # percentage
    occupied_rooms: int
    total_rooms: int


class RoomStatusDistribution(BaseModel):
    """Room status distribution"""
    available: int
    occupied: int
    cleaning: int
    reserved: int
    out_of_service: int


class OccupancyReportResponse(BaseModel):
    """Occupancy report response"""
    occupancy_rate: float  # Overall percentage
    total_rooms: int
    occupied_rooms: int
    available_rooms: int
    room_status_distribution: RoomStatusDistribution
    by_period: List[OccupancyByPeriod]  # Chart data
    start_date: date
    end_date: date


# ============================================================================
# Booking Report Schemas
# ============================================================================

class BookingByPeriod(BaseModel):
    """Booking data point for chart"""
    period: str  # "2025-01"
    total_bookings: int
    confirmed: int
    cancelled: int
    checked_in: int


class BookingReportResponse(BaseModel):
    """Booking report response"""
    total_bookings: int
    confirmed_bookings: int
    cancelled_bookings: int
    checked_in_bookings: int
    cancellation_rate: float  # percentage
    conversion_rate: float  # (checked_in / confirmed) * 100
    total_deposit: float
    by_period: List[BookingByPeriod]  # Chart data
    start_date: date
    end_date: date


# ============================================================================
# Customer Report Schemas
# ============================================================================

class TopCustomer(BaseModel):
    """Top customer data"""
    customer_id: int
    full_name: str
    phone_number: str
    total_spending: float
    visit_count: int
    last_visit: Optional[datetime]


class CustomerReportResponse(BaseModel):
    """Customer report response"""
    top_customers: List[TopCustomer]
    total_customers: int
    new_customers: int  # Customers with first visit in date range
    returning_customers: int  # Customers with > 1 visit


# ============================================================================
# Summary Report Schemas
# ============================================================================

class QuickStat(BaseModel):
    """Quick stat for dashboard"""
    label: str
    value: str
    change: Optional[float] = None  # percentage change vs previous period
    trend: Optional[str] = None  # "up" | "down" | "neutral"


class SummaryReportResponse(BaseModel):
    """Summary report for dashboard overview"""
    # Revenue
    total_revenue: float
    revenue_vs_previous: Optional[float] = None

    # Occupancy
    occupancy_rate: float
    occupancy_vs_previous: Optional[float] = None

    # Check-ins/outs
    total_checkins: int
    total_checkouts: int

    # Bookings
    total_bookings: int
    bookings_vs_previous: Optional[float] = None

    # Customers
    total_customers: int
    new_customers: int

    # Quick stats for cards
    quick_stats: List[QuickStat]

    # Date range
    start_date: date
    end_date: date

File: frontend/src/api/reports.ts
/**
 * Reports API Client (Phase 8)
 * API calls for reports and analytics
 */
import apiClient from './client'

// ============================================================================
// Types
// ============================================================================

export interface RevenueByPeriod {
  period: string
  revenue: number
  count: number
}

export interface RevenueReport {
  total_revenue: number
  total_transactions: number
  average_transaction: number
  by_payment_method: Record<string, number>
  by_stay_type: Record<string, number>
  by_period: RevenueByPeriod[]
  start_date: string
  end_date: string
}

export interface OccupancyByPeriod {
  period: string
  occupancy_rate: number
  occupied_rooms: number
  total_rooms: number
}

export interface RoomStatusDistribution {
  available: number
  occupied: number
  cleaning: number
  reserved: number
  out_of_service: number
}

export interface OccupancyReport {
  occupancy_rate: number
  total_rooms: number
  occupied_rooms: number
  available_rooms: number
  room_status_distribution: RoomStatusDistribution
  by_period: OccupancyByPeriod[]
  start_date: string
  end_date: string
}

export interface BookingByPeriod {
  period: string
  total_bookings: number
  confirmed: number
  cancelled: number
  checked_in: number
}

export interface BookingReport {
  total_bookings: number
  confirmed_bookings: number
  cancelled_bookings: number
  checked_in_bookings: number
  cancellation_rate: number
  conversion_rate: number
  total_deposit: number
  by_period: BookingByPeriod[]
  start_date: string
  end_date: string
}

export interface TopCustomer {
  customer_id: number
  full_name: string
  phone_number: string
  total_spending: number
  visit_count: number
  last_visit: string | null
}

export interface CustomerReport {
  top_customers: TopCustomer[]
  total_customers: number
  new_customers: number
  returning_customers: number
}

export interface QuickStat {
  label: string
  value: string
  change: number | null
  trend: 'up' | 'down' | 'neutral' | null
}

export interface SummaryReport {
  total_revenue: number
  revenue_vs_previous: number | null
  occupancy_rate: number
  occupancy_vs_previous: number | null
  total_checkins: number
  total_checkouts: number
  total_bookings: number
  bookings_vs_previous: number | null
  total_customers: number
  new_customers: number
  quick_stats: QuickStat[]
  start_date: string
  end_date: string
}

export interface CheckInListItem {
  id: number
  room_number: string
  room_type_name: string
  customer_name: string
  customer_phone: string
  stay_type: 'overnight' | 'temporary'
  check_in_time: string
  expected_check_out_time: string | null
  check_out_time: string | null
  total_amount: number
  payment_method: string
  status: 'checked_in' | 'checked_out'
  number_of_nights: number | null
  number_of_guests: number
}

export interface CheckInsListResponse {
  check_ins: CheckInListItem[]
  total_count: number
  total_revenue: number
  start_date: string
  end_date: string
}

// ============================================================================
// API Functions
// ============================================================================

export const reportsApi = {
  /**
   * Get revenue report
   */
  async getRevenueReport(
    startDate: string,
    endDate: string,
    groupBy: 'day' | 'month' = 'day'
  ): Promise<RevenueReport> {
    const response = await apiClient.get<RevenueReport>('/reports/revenue', {
      params: {
        start_date: startDate,
        end_date: endDate,
        group_by: groupBy
      }
    })
    return response.data
  },

  /**
   * Get occupancy report
   */
  async getOccupancyReport(
    startDate: string,
    endDate: string
  ): Promise<OccupancyReport> {
    const response = await apiClient.get<OccupancyReport>('/reports/occupancy', {
      params: {
        start_date: startDate,
        end_date: endDate
      }
    })
    return response.data
  },

  /**
   * Get booking report
   */
  async getBookingReport(
    startDate: string,
    endDate: string
  ): Promise<BookingReport> {
    const response = await apiClient.get<BookingReport>('/reports/bookings', {
      params: {
        start_date: startDate,
        end_date: endDate
      }
    })
    return response.data
  },

  /**
   * Get customer report
   */
  async getCustomerReport(limit: number = 10): Promise<CustomerReport> {
    const response = await apiClient.get<CustomerReport>('/reports/customers', {
      params: { limit }
    })
    return response.data
  },

  /**
   * Get summary report (for dashboard)
   */
  async getSummaryReport(
    startDate?: string,
    endDate?: string
  ): Promise<SummaryReport> {
    const response = await apiClient.get<SummaryReport>('/reports/summary', {
      params: {
        start_date: startDate,
        end_date: endDate
      }
    })
    return response.data
  },

  /**
   * Get check-ins list for reports table
   */
  async getCheckInsList(
    startDate: string,
    endDate: string
  ): Promise<CheckInsListResponse> {
    const response = await apiClient.get<CheckInsListResponse>('/reports/check-ins', {
      params: {
        start_date: startDate,
        end_date: endDate
      }
    })
    return response.data
  }
}
